Use 2 Redis instances in Server image, redis-cache & redis-store

mrnugget opened the pull request:
This fixes #5015 by changing the Server image configuration to launch two Redis instances, one to be used as a cache and one as a store.

The cache instance runs on port 6380 (unused according to [Wikipedia](https://en.wikipedia.org/wiki/List_of_TCP_and_UDP_port_numbers)) and has `maxmemory-policy allkeys-lru`, which allows us to get rid of TTLs on the cache keys (which I want to do as a separate follow-up PR).

After looking at the usage of Redis on my local Sourcegraph instance, I decided to give the cache instance 800mb of memory and the store instance 200mb (1GB was the maxmemory value used before). If someone has better suggestions for this ratio, let me know.

I built and ran the server image locally and tested that both Redis instances get used by `docker exec`ing into the Server image and running `redis-cli -p <port> MONITOR` two times. See:

![double_redis](https://user-images.githubusercontent.com/1185253/62293427-e1d5e280-b468-11e9-9ebe-a237584b8aaf.gif)

---

**FINAL SUMMARY**:

* This adds a new Redis instance to the Docker server image, `redis-cache`, that uses `allkeys-lru` eviction policy.
* The `rcache` package now solely uses this `redis-cache` instance
* Existing non-cache data in Redis is migrated to the non-cache Redis instance, `redis-store`.
* Existing cache data, prefixed with the previous `rcacheDataVersion` of `v1`, will be deleted in the background when frontend boots up — this happens for all deployments of Sourcegraph.
* A warning message is printed when Sourcegraph boots up and detects that the configured `redis-cache` instance doesn't use an `allkeys-lru` eviction policy — this happens for all deployments of Sourcegraph.

nicksnyder commented:
Thanks for following up on this!

> After looking at the usage of Redis on my local Sourcegraph instance, I decided to give the cache instance 800mb of memory and the store instance 200mb (1GB was the maxmemory value used before). If someone has better suggestions for this ratio, let me know.

redis-store has session data (which is more important), so for large deployments I would be worried that this isn't the right split. If we just make them both 1GB then overall memory usage would increase but neither use case would risk regression.

> which allows us to get rid of TTLs on the cache keys

For customers that have configured REDIS_STORE_ENDPOINT and/or REDIS_CACHE_ENDPOINT, we need to ensure that they have correct configurations as well. Is there any way to query/validate redis config at runtime? If so, we can check from within the app and terminate with an error message. If not, we need to proactively communicate with existing customers who are using these env vars, and add/update documentation for future customers who use these.

This raises another (separate) question for me: when is session data in redis-store evicted? If not globally, does each session key have an expiration within redis?

mrnugget commented:
> redis-store has session data (which is more important), so for large deployments I would be worried that this isn't the right split. If we just make them both 1GB then overall memory usage would increase but neither use case would risk regression. What do you think?

Yeah, sounds good to me. Eliminates the risk of getting the ratio wrong and is consistent.



> For customers that have configured REDIS_STORE_ENDPOINT and/or REDIS_CACHE_ENDPOINT, we need to ensure that they have correct configurations as well. Is there any way to query/validate redis config at runtime? If so, we can check from within the app and terminate with an error message. If not, we need to proactively communicate with existing customers who are using these env vars, and add/update documentation for future customers who use these.

Yeah, we can check that:

```
$ redis-cli CONFIG GET maxmemory-policy
1) "maxmemory-policy"
2) "noeviction"
```

I'll add the check. Note: this will make this a *breaking change* that we'll need to document/communicate appropriately.

> This raises another (separate) question for me: when is session data in redis-store evicted? If not globally, does each session key have an expiration within redis?

Yeah, they do. The underlying `redistore` library uses 30 days as the default expiration time for session data/cookies: https://sourcegraph.com/github.com/boj/redistore/-/blob/redistore.go#L24

---

Do we need to migrate existing data?

I don't think we need to migrate the caches, but what about the session data, usage stats and license user counts? (Those 3 [seem to be the only users of `redispool.Store`](https://sourcegraph.sgdev.org/search?q=repo:%5Esourcegraph/sourcegraph%24+redispool.Store))

If so, the code in this PR would need to change so that
1. the `redis-store` uses the same data directory as before
2. the `redis-store` gets a migration that deletes all keys beginning with `rcache`'s `globalPrefix` (defined here: https://sourcegraph.sgdev.org/sourcegraph/sourcegraph@1ed7c5fbd6300aff6436478d1ebd91d02764c8e7/-/blob/pkg/rcache/rcache.go#L20).

Notes regarding the migration:
* **We can only run this if `redis-store` and `redis-cache` point to two different instances**. We should compare the addresses and the _actual instance we're connected to_ (since two addresses might point to the same instance). We can use Redis' `INFO` to get the `run_id` and compare that:
    ```bash
    $ redis-cli INFO server | grep run_id
    run_id:542c0cda6ed45c9ebc58e87520ef3bfaa2d7d508
    ```
* We should probably use this occasion to increase `rcache`'s `dataVersion` (defined [here](https://sourcegraph.sgdev.org/sourcegraph/sourcegraph@1ed7c5fbd6300aff6436478d1ebd91d02764c8e7/-/blob/pkg/rcache/rcache.go#L20)), which serves as the `globalPrefix` from `v1` to `v2`. With that, we can run the "migration" every time we connect to `redis-store` without deleting current cache data.
* We should only run the migration in server image deployments. Best place would probably be in `frontend`. We check whether `DEPLOY_TYPE` is `docker-container`, check whether `redis-store` and `redis-cache` point to two different instances and if so, we scan all keys starting with `v1` and delete them.

Opinions?

Of course, if we don't need to migrate things, then we could merge the PR as it is.

keegancsmith commented:
On further thought it might make sense to migrate the current redis in an instance to redis-store, to avoid us losing all session and other data we store in there.

mrnugget commented:
> On further thought it might make sense to migrate the current redis in an instance to redis-store, to avoid us losing all session and other data we store in there.

So you agree with the plan in the second part of [my comment](https://github.com/sourcegraph/sourcegraph/pull/5040#issuecomment-517604644)?

tsenart commented:
> So you agree with the plan in the second part of my comment?

Looks good to me. On another note, since we're touching the eviction policy, have you looked at https://redis.io/topics/lru-cache#the-new-lfu-mode?

keegancsmith commented:
Sorry missed this discussion :)

> > redis-store has session data (which is more important), so for large deployments I would be worried that this isn't the right split. If we just make them both 1GB then overall memory usage would increase but neither use case would risk regression. What do you think?
> 
> Yeah, sounds good to me. Eliminates the risk of getting the ratio wrong and is consistent.

Also agree.

> > For customers that have configured REDIS_STORE_ENDPOINT and/or REDIS_CACHE_ENDPOINT, we need to ensure that they have correct configurations as well. Is there any way to query/validate redis config at runtime? If so, we can check from within the app and terminate with an error message. If not, we need to proactively communicate with existing customers who are using these env vars, and add/update documentation for future customers who use these.
> 
> Yeah, we can check that:
> 
> ```
> $ redis-cli CONFIG GET maxmemory-policy
> 1) "maxmemory-policy"
> 2) "noeviction"
> ```
> 
> I'll add the check. Note: this will make this a _breaking change_ that we'll need to document/communicate appropriately.

You can print a loud warning to avoid making it a breaking change. Additionally, if store and cache data is in the same instance then we should allow it. This tells me your other PR which always includes a TTL is still important for this use case.

I don't think we should prevent people from doing this. Just think how long we have been using a single instance without any complaints related to the issue we ran into.

> Do we need to migrate existing data?

Yes, but only store.

> If so, the code in this PR would need to change so that
> 
> 1. the `redis-store` uses the same data directory as before
> 2. the `redis-store` gets a migration that deletes all keys beginning with `rcache`'s `globalPrefix` (defined here: https://sourcegraph.sgdev.org/sourcegraph/sourcegraph@1ed7c5fbd6300aff6436478d1ebd91d02764c8e7/-/blob/pkg/rcache/rcache.go#L20).

SGTM


> Notes regarding the migration:
> 
> * **We can only run this if `redis-store` and `redis-cache` point to two different instances**. We should compare the addresses and the _actual instance we're connected to_ (since two addresses might point to the same instance). We can use Redis' `INFO` to get the `run_id` and compare that:
>   ```shell
>   $ redis-cli INFO server | grep run_id
>   run_id:542c0cda6ed45c9ebc58e87520ef3bfaa2d7d508
>   ```

We should only do the migration if we are starting up redis. IE the user hasn't set any of the redis env vars. Then we don't need to worry about this. If a user has there own redis then there isn't any migration we can do.

> * We should probably use this occasion to increase `rcache`'s `dataVersion` (defined [here](https://sourcegraph.sgdev.org/sourcegraph/sourcegraph@1ed7c5fbd6300aff6436478d1ebd91d02764c8e7/-/blob/pkg/rcache/rcache.go#L20)), which serves as the `globalPrefix` from `v1` to `v2`. With that, we can run the "migration" every time we connect to `redis-store` without deleting current cache data.

Good idea. Then for both redis-store and redis-cache we can drop those.

> * We should only run the migration in server image deployments. Best place would probably be in `frontend`. We check whether `DEPLOY_TYPE` is `docker-container`, check whether `redis-store` and `redis-cache` point to two different instances and if so, we scan all keys starting with `v1` and delete them.

We could do the migration in the server binary.

A new point. We should double check if we use a good config in deploy-sourcegraph.

nicksnyder commented:
> > Do we need to migrate existing data?

> Yes, but only store.

Agree.

> I don't think we should prevent people from doing this. Just think how long we have been using a single instance without any complaints related to the issue we ran into.

I think we should discourage (ideally prevent) new instances from doing this. It seems like it will just lead to hard-to-diagnose ephemeral issues down the line if the store does fill up temporarily. I don't have a problem with back-compat for current instances. Maybe a loud warning in the logs is sufficient?

mrnugget commented:
I just occurred to me I might have missed how simplified the migration is by increasing rcache's `dataVersion` from `v1` to `v2`. I assumed that we need to do a lot of these checks (did we boot up redis? are these different instances?), but if we keep the data in `redis-store` anyway and we just add another cache instance that uses a new cache key... All we need to do on boot is delete the old cache data under the `v1` prefix.

That could/should even happen in non-server-image deployments, otherwise we'd have stale cache data without TTLs lying around.

Or am I missing something now?

keegancsmith commented:
> I just occurred to me I might have missed how simplified the migration is by increasing rcache's `dataVersion` from `v1` to `v2`. I assumed that we need to do a lot of these checks (did we boot up redis? are these different instances?), but if we keep the data in `redis-store` anyway and we just add another cache instance that uses a new cache key... All we need to do on boot is delete the old cache data under the `v1` prefix.
> 
> That could/should even happen in non-server-image deployments, otherwise we'd have stale cache data without TTLs lying around.
> 
> Or am I missing something now?

Nope I don't think you are missing anything. Sounds good.

mrnugget commented:
@nicksnyder @keegancsmith I'm done. I've added the "delete outdated cache" migration, changed the `redis-store` data dir to use the old one (which keeps the data around) and added a warning message. Can you re-review the code in here?

@tsenart I haven't looked at that eviction policy yet, no. I'll take a look this week and if it fits, I'll follow up on this :)

keegancsmith commented on cmd/frontend/internal/bg/delete_old_cache_data_in_redis.go:
```diff
@@ -0,0 +1,116 @@
+	// Only run on separate instances
+	connsToPurge := make(map[string]redis.Conn)
+	connsToPurge[storeRunID] = storeConn
+	connsToPurge[cacheRunID] = cacheConn
```
minor: we could just simplify this and always run the purge twice? Is it quick if there is nothing to purge? If it is, I'd just do that.

mrnugget replied:
Done!

keegancsmith commented on cmd/frontend/internal/bg/delete_old_cache_data_in_redis.go:
```diff
@@ -0,0 +1,116 @@
+	iter := 0
+	keys := make([]string, 0)
+	for {
+		arr, err := redis.Values(c.Do("SCAN", iter, "MATCH", pattern))
```
we have code in our testing package which deletes all keys matching a prefix. It does it via the redis lua script stuff and is quite efficient. I'd consider using that instead since its much less code and more efficient.

mrnugget replied:
Done!

keegancsmith commented on pkg/rcache/rcache.go:
```diff
@@ -17,7 +17,7 @@ import (
 // effectively resetting the cache at the same time the new code
 // is deployed.
-const dataVersion = "v1"
+const dataVersion = "v2"
```
minor: Maybe the migration which deletes older versions should live in this package, and the `bg` package then just calls this one. WDYT?

mrnugget replied:
Good point. Agreed. Will do this.

mrnugget commented:
Incorporated the suggestions from @keegancsmith. Tested with local build of server image on top of data directory created and populated by `3.6.0`. After booting up the new image with 2 instances:

```bash
$ docker exec -it 03a0255ca03a bash
bash-4.4# redis-cli -p 6379 KEYS '*'
1) "user_activity:1"
2) "session_W6P5ARC223EKSROUCUZAWZTXI74D2M35M7YVFBJSMYZDPBWPX73Q"
3) "user_activity::usersactive:2019-08-06"
4) "user_activity:searchoccurred"
5) "user_activity::usersactive:2019-08-05"
bash-4.4# redis-cli -p 6380 KEYS '*'
1) "v2:gh_repo:uefipg9tCJgR_Em3xlzqBUnBhZNcZ5H_Js_cA4mjKCY=:1:mdewoljlcg9zaxrvcnkxmdq2ndu2nji="
2) "v2:gh_repo:uefipg9tCJgR_Em3xlzqBUnBhZNcZ5H_Js_cA4mjKCY=:0:mrnugget/emacs.d"
3) "v2:http:http://127.0.0.1:3180/search/repositories?page=1&per_page=100&q=org%3Amrnugget+emacs"
```

nicksnyder commented:
I don't need to review this, but please update the PR description to reflect the end state.

mrnugget commented:
> I don't need to review this, but please update the PR description to reflect the end state.

Done!
